red4ext/src/downcast.rs

```diff
@@ -0,0 +1,172 @@
+//!
+//! #[derive(Default, Clone)]
+//! #[repr(transparent)]
+//! pub struct System(Ref<IScriptable>);
+//!
+//! unsafe impl RefRepr for System {
+//!     type Type = Strong;
+//!     const CLASS_NAME: &'static str = "My.System";
+//! }
+//!
+//! #[redscript_import]
+//! impl System {
+//!     /// import custom callback constructor
+//!     fn create_custom_callback(&self, message: RedString) -> CustomCallback;
+//! }
+//!
+//! #[derive(Default, Clone)]
+//! #[repr(transparent)]
+//! pub struct CustomCallback(Ref<IScriptable>);
+//!
+//! unsafe impl RefRepr for CustomCallback {
+//!     type Type = Strong;
+//!     const CLASS_NAME: &'static str = "My.CustomCallback";
+//! }
+//!
+//! /// downcast trait and friends make it more convenient,
+//! /// ⚠️ but you MUST make sure `CustomCallback` is indeed a child class of `DelayCallback`.
+//! unsafe impl IntoTypedRef<DelayCallback> for CustomCallback {
+//!     fn into_typed_ref(self) -> TypedRef<DelayCallback> {
+//!         TypedRef::new(self.0)
+//!     }
+//! }
+//!
+//! impl System {
+//!     /// then, whenever you need to trigger your callback
+//!     fn when_something_happens(&self) {
+//!         let callback: CustomCallback = self.create_custom_callback(RedString::new("Hello from My.System!"));
+//!         // ✅ now simply `downcast` your child class
+//!         self.delay_callback_next_frame(callback.downcast());
+//!     }
+//! }
+//! ```
+
+use std::marker::PhantomData;
+
+use super::conv::RefRepr;
+use super::ffi::IScriptable;
+use super::types::Ref;
+
+#[derive(Default, Clone)]
+#[repr(transparent)]
+pub struct TypedRef<T>(Ref<IScriptable>, PhantomData<T>);
+
+impl<T> TypedRef<T> {
+    pub fn new(reference: Ref<IScriptable>) -> Self {
+        Self(reference, PhantomData)
+    }
+
+    pub fn into_inner(self) -> Ref<IScriptable> {
+        self.0
+    }
+}
+
+/// # Safety
+///
+/// implementations of this trait are only valid if your implementors are indeed child classes of `Parent`
+pub unsafe trait IntoTypedRef<Parent: RefRepr + Default + Clone> {
+    /// cast a reference into a parent typed reference
+    fn into_typed_ref(self) -> TypedRef<Parent>;
+}
+
+pub trait FromTypedRef<Parent: RefRepr + Default + Clone>: Sized {
+    /// cast any self typed reference into itself
+    fn from_typed_ref(reference: TypedRef<Parent>) -> Self;
+}
+
+pub trait Downcast<Parent>: IntoTypedRef<Parent>
+where
+    Self: IntoTypedRef<Parent>,
+    Parent: RefRepr + Default + Clone + FromTypedRef<Parent>,
+{
+    /// automatically downcast a wrapper over `Ref<IScriptable>`
+    /// into its parent
+    fn downcast(self) -> Parent
+    where
+        Self: Sized,
+    {
+        Parent::from_typed_ref(self.into_typed_ref())
+    }
+}
+
+/// automatically implements downcast for any child classes that can be turned into their parent class
+impl<Child, Parent> Downcast<Parent> for Child
+where
+    Self: IntoTypedRef<Parent>,
+    Parent: RefRepr + Default + Clone + FromTypedRef<Parent>,
+{
+}
```

Comment:
I'm not sure what the purpose of `TypedRef<T>` here is,
I'd do it differently, but maybe I'm missing something:
If we need downcasting, we can just use a marker trait with an associated type because redscript only allows single inheritance:
```rs
unsafe trait AsBaseClass: Sized {
    type BaseClass;
}

fn as_base<A: AsBaseClass>(a: &A) -> &A::BaseClass {
    unsafe { std::mem::transmute(a) }
}
```